### PR TITLE
Custom exceptions for the vulnerability consumer

### PR DESCRIPTION
**Pull request by mir-am**
## Description
In this PR, custom exceptions will be implemented for the vulnerability consumer. The required custom exceptions are described in issue #249.

## Motivation and context
To better measure the precision and performance of the vulnerability consumer, custom exceptions are needed.

## Testing
We need to run the plugin again and analyze its output Kafka topic, and also its error topic.

## Task list  
- [] Implement classes for custom exceptions described in #249 
- [] Throw custom exceptions in the method `injectVulnerabilityIntoDB` in the specified places

### CODE REVIEW

**Comment by proksch** [comment]
Not sure about the state of this PR... it is quite old and even back then, some checks did not pass. Now we even have conflicts on top. Is this change still relevant for us?

File: analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/VulnerabilityConsumer.java [hunk 1]
```diff
@@ -160,48 +160,65 @@ public long getMaxConsumeTimeout() {
...
             var vulnCallableIds = new HashSet<Long>();
             var vulnPkgVersionIds = new HashSet<>(pkgVersionIds);
-            fastenUris.forEach(uri -> {
-                var callIds = metadataUtility.getCallableIdsForFastenUri(uri, vulnPkgVersionIds, context);
-                callIds.forEach(id -> fullFastenUris.add(metadataUtility.getFullFastenUri(uri, id)));
-                vulnCallableIds.addAll(callIds);
-            });
 
-            v.setFastenUris(fullFastenUris);
-            writePatchDate(v);
```
Comment (amottier) [inline analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/VulnerabilityConsumer.java:204]: It seems that this call to `writePatchDate(v);` was removed. I'm not sure it was deliberate?

File: analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/VulnerabilityConsumer.java [hunk 2]
```diff
@@ -257,6 +274,7 @@ public static void writePatchDate(Vulnerability v) {
                 if (set)
                     v.setPatchDate(sdf.format(latestDate));
             } catch (Exception e) {
+                // TODO: Why an empty exception block?
```
Comment (amottier) [inline analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/VulnerabilityConsumer.java:277]: I can't explain it but I would suggest to remove it and implement a better error handling once we figure out which exceptions were silently catch.

File: analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/db/MetadataUtility.java [hunk 3]
```diff
@@ -244,8 +244,23 @@ public Long getPackageId(String pkgName, DSLContext context) {
                 .fetch();
 
         var callIds = new HashSet<>(callInfo.map(Record2::value1));
+//        callIds.addAll(callIdsSubProj);
+        callInfo.forEach(r -> cache.callIdToPkgVsn.put(r.value1(), cache.pkgVsnIdToVsn.get(r.value2())));
+        return callIds;
+    }
```
Comment (amottier) [inline analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/db/MetadataUtility.java:250]: Modification in this file does not seems to be related to exceptions handling.
I would suggest to isolate them in a different PR if they are still relevant.

File: analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/utils/exceptions/CallablesNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package eu.fasten.analyzer.vulnerabilityconsumer.utils.exceptions;
+
+/**
+ * Exception thrown if patches are present but no fasten_uri is found for the vulnerability.
+ */
+public class CallablesNotFoundException extends Exception {
+    /**
+     * Default constructor without a message.
+     */
+    public CallablesNotFoundException() {
```
Comment (amottier) [inline analyzer/vulnerability-consumer/src/main/java/eu/fasten/analyzer/vulnerabilityconsumer/utils/exceptions/CallablesNotFoundException.java:10]: I would suggest to add a little bit of context to this exception such as information about the callable that we are searching for.

Same apply to other custom Exception classes.

**Review by amottier — COMMENTED** [review]
Overall PR looks quite good to me and only require minor adjustment.
I think we can still benefit from those more specific exceptions types but @mir-am I let you decide if you want to keep it or drop it.

**Comment by mir-am** [comment]
@amottier
Thanks for reviewing the PR.
The changes in this PR probably wouldn't be used as the vulnerability consumer is now replaced by `vulnerability-statement-processor`. At some point, we'll remove this plug-in from the project. (See #316)

**Comment by amottier** [comment]
@mir-am is it ok to close this PR and remove associate branch as #316 is now merged?

**Comment by mir-am** [comment]
> @mir-am is it ok to close this PR and remove associate branch as #316 is now merged?

Yes. Thanks.

**Pull request closed without merging**